compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -275,7 +282,15 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: LocalDefId) -> ty::EarlyBinder<'_
                 let args = ty::GenericArgs::identity_for_item(tcx, def_id);
                 Ty::new_fn_def(tcx, def_id.to_def_id(), args)
             }
-            ForeignItemKind::Static(t, _, _) => icx.lower_ty(t),
+            ForeignItemKind::Static(ty, _, _) => {
+                let ty = icx.lower_ty(ty);
+                // MIR relies on references to statics being scalars.
+                // Verify that here to avoid ill-formed MIR.
+                match check_static_item(tcx, def_id, ty, false) {
```

Comment:
```suggestion
                // Verify that here to avoid ill-formed MIR.
                // We skip the `Sync` check to avoid cycles for type-alias-impl-trait,
                // relying on the fact that non-Sync statics don't ICE the rest of the compiler.
                match check_static_item(tcx, def_id, ty, /* should_check_for_sync */ false) {
```